test/conftest.py

```diff
@@ -3,8 +3,14 @@
 import pytest
 
 store_loc = str(os.path.join("/tmp/", str(uuid.uuid4())))
+redis_port_num = 6379
```

Comment:
It will not automatically find free ports, but it `redis-server` does exit with error if another instance of `redis-server` is already running on that port. Incidentally, running `nc -l localhost 1234` does not cause `redis-server --port 1234` to crash, so it looks like it's only straightforward to detect (and recover from) port collisions with multiple Redis instances.

Since we're going to expose Redis configuration (outside of the default port 6379) in the config file, I think the best option here is to detect collisions and gracefully shut down, rather than try to find an open port. It would be difficult to reason about the behavior of the system if the user asked us to start up on a certain port and we ended up choosing another because the desired port was busy.

Reply:
This all makes sense. I’m just thinking that it would be preferable to _not_ require users to specify ports if they don’t have a reason to care. I think recovering from error is a good idea, but even better if we can automate and inform so the rest of the system knows. Could we start by trying 6379, warn if that fails, then increment until we succeed?

Reply:
So, it sounds like we’re talking about two options right now:

1. Require users to specify ports of Redis instance(s) in config files. This lets us be sure that we won’t have any collisions (that the user isn’t themselves responsible for, at least) in the system, but it requires the user to configure the ports. If we attempt to gracefully recover from a fault here, then we’ll be running the system in a configuration that wasn’t specified by the user. However, the user would be able to rectify this error with an additional configuration, and the system would respond deterministically.
2. Do not require users to specify ports of Redis instances, and instead calculate them on the fly. This makes it easier for users to run the system, but they won’t be able to reason about the ports that Redis is running on in advance, which I’m not sure if we really have a use case for, either at the moment or on the horizon.

Perhaps we could combine the two and allow improv to find a free port for a single instance of Redis out of the box, but if a port *is* specified for a more advanced configuration, we’ll just throw the error if there’s a collision. In some sense, if the user is knowledgeable enough about the system to have a reason to specify a port, they should be expected to be able to fix the error if the port they specify is already in use. This way we get easy out of the box behavior, but the behavior of the store is still deterministic when required. Thoughts?

Reply:
Yes, exactly. An explicit configuration should take precedence and throw and error if the port is in use. We should not fail silently.

If users are not used to thinking about ports or don't need to know, it's best not to require them to choose one. In that case, using the Redis default should be fine 99% of the time, but if it happens not to be, I'd prefer to move on to the next available one and not require users to dig in.

And of course this requires that actors be informed about which port Redis is actually on. We should never assume the config file is the truth.